up-streamer/src/ustreamer.rs

```diff
@@ -0,0 +1,973 @@
+                        UCode::NOT_FOUND,
+                        "No in_transport and out_authority found corresponding",
+                    );
+                    error!(
+                        "{}:{}:{} Deleting forwarding rule failed: {:?}",
+                        self.name,
+                        USTREAMER_TAG,
+                        USTREAMER_FN_DELETE_FORWARDING_RULE_TAG,
+                        &no_in_transport_out_authority_found
+                    );
+                    Err(no_in_transport_out_authority_found)
+                }
+            } else {
+                Ok(())
+            }
+        } else {
+            let err = Err(UStatus::fail_with_code(
+                UCode::NOT_FOUND,
+                format!(
+                    "{} is not routed. No rule to delete.",
+                    Self::forwarding_id(&r#in, &out)
+                ),
+            ));
+            error!(
+                "{}:{}:{} Deleting forwarding rule failed: {:?}",
+                self.name, USTREAMER_TAG, USTREAMER_FN_DELETE_FORWARDING_RULE_TAG, err
+            );
+            err
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct ComparableTransport {
+    transport: Arc<Box<dyn UTransport>>,
+}
+
+impl ComparableTransport {
+    pub fn new(transport: Arc<Box<dyn UTransport>>) -> Self {
+        Self { transport }
+    }
+}
+
+impl Hash for ComparableTransport {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Arc::as_ptr(&self.transport).hash(state);
+    }
+}
+
+impl PartialEq for ComparableTransport {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.transport, &other.transport)
+    }
+}
+
+impl Eq for ComparableTransport {}
+
+const TRANSPORT_FORWARDER_TAG: &str = "TransportForwarder:";
+const TRANSPORT_FORWARDER_FN_MESSAGE_FORWARDING_LOOP_TAG: &str = "message_forwarding_loop():";
+pub(crate) struct TransportForwarder {}
+
+impl TransportForwarder {
+    async fn new(
+        out_transport: Arc<Box<dyn UTransport>>,
+        message_receiver: Receiver<Arc<UMessage>>,
+    ) -> Self {
+        let out_transport_clone = out_transport.clone();
+        let message_receiver_clone = message_receiver.clone();
+        task::spawn(Self::message_forwarding_loop(
+            UUIDBuilder::build().to_hyphenated_string(),
+            out_transport_clone,
+            message_receiver_clone,
+        ));
+
+        Self {}
+    }
+
+    async fn message_forwarding_loop(
+        id: String,
+        out_transport: Arc<Box<dyn UTransport>>,
+        message_receiver: Receiver<Arc<UMessage>>,
+    ) {
+        while let Ok(msg) = message_receiver.recv().await {
+            debug!(
+                "{}:{}:{} Attempting send of message: {:?}",
+                id,
+                TRANSPORT_FORWARDER_TAG,
+                TRANSPORT_FORWARDER_FN_MESSAGE_FORWARDING_LOOP_TAG,
+                msg
+            );
+
+            let send_res = out_transport.send(msg.deref().clone()).await;
```

Comment:
That may be nice... yes, unfortunately we end up having to clone the `UMessage` fairly often with the current UTransport::send() API.

Would we want to have 2 x send APIs (w/ different names obviously) on UTransport, then? One accepting Arc<UMessage>, the other accepting UMessage?

Not great if an implementor of `UTransport` has to do both. I suppose it could boil down to the one taking UMessage having a default impl of putting the UMessage in an Arc then calling the Arc<UMessage> one?

Reply:
I think this needed further follow-up in up-rust.